pitop/robotics/pan_tilt_controller.py

```diff
@@ -1,12 +1,143 @@
-from pitop import ServoMotor
-from pitop.system.port_manager import PortManager
+import configparser
+from os.path import (
+    exists,
+    isdir,
+    join,
+)
+from pathlib import Path
 
+from pitop.core.exceptions import UninitializedComponent
+from pitop.core.mixins import (
+    Stateful,
+    Recreatable,
+)
+from pitop.pma import ServoMotor
 
-class PanTiltController:
 
-    def __init__(self, servo_pan_port="S0", servo_tilt_port="S3"):
+class PanTiltController(Stateful, Recreatable):
+    CALIBRATION_FILE_DIR = ".config/pi-top/sdk"
+    CALIBRATION_FILE_NAME = "alex.conf"
+    _initialized = False
+    _pan_servo = None
+    _tilt_servo = None
+
+    def __init__(self, servo_pan_port="S0", servo_tilt_port="S3", name="pan_tilt"):
+        self.name = name
         self._pan_servo = ServoMotor(servo_pan_port)
         self._tilt_servo = ServoMotor(servo_tilt_port)
-        self.__port_manager = PortManager()
-        self.__port_manager.register_pma_component(self._pan_servo)
-        self.__port_manager.register_pma_component(self._tilt_servo)
+        self.__calibration_file_path = join(str(Path.home()), self.CALIBRATION_FILE_DIR, self.CALIBRATION_FILE_NAME)
+        self._initialized = True
+
+        Stateful.__init__(self, children=['_pan_servo', '_tilt_servo'])
+        Recreatable.__init__(self, config_dict={'servo_pan_port': servo_pan_port, 'servo_tilt_port': servo_tilt_port, 'name': name})
+
+    def is_initialized(fcn):
+        def check_initialization(self, *args, **kwargs):
+            if not self._initialized:
+                raise UninitializedComponent("PanTiltController not initialized")
+            return fcn(self, *args, **kwargs)
+        return check_initialization
+
+    @property
+    @is_initialized
+    def pan_servo(self):
+        return self._pan_servo
+
+    @property
+    @is_initialized
+    def tilt_servo(self):
+        return self._tilt_servo
+
+    @is_initialized
+    def calibrate(self, save=True, reset=False):
```

Comment:
Whilst this does what it is supposed to, it's not very intuitive and lacks sophistication.

I wonder if we can/should create some sort of "pan-tilt calibrator", or similar that uses the miniscreen and buttons to simplify the process. It could be a separate program that is run - that is, part of the CLI - or something that can be called directly from Python. It would be great if a pi-top object could have its various subsystems calibrated, such as:
* iteratively tweaking a camera's resolution to optimise performance
* automatically determine tilt zero point by using camera as a sort of 'spirit level'
etc.

Just thinking out loud here, but I would love to see this improved.

Reply:
Just created https://github.com/pi-top/pi-top-Python-SDK/issues/267 with your suggestions so we can work on it later on